scapy/layers/dns.py

```diff
@@ -88,6 +88,30 @@ def dns_get_str(s, p, pkt=None, _internal=False):
     return name, p, s[burned:]
 
 
+def dns_encode(x, check_built=False):
+    """Encodes a bytes string into the DNS format
+
+    :param x: the string
+    :param check_built: detect already-built strings and ignore them
+    :returns: the encoded bytes string
+    """
+    if not x or x == b".":
+        return b"\x00"
+
+    built = check_built
+    built &= b"." not in x
+    built &= ((x[-1] == b"\x00") or ((orb(x[-2]) & 0xc0)) == 0xc0)
```

Comment:
This bit doesn't seem to be needed though. Due to the final dot added by any2i, the presence or absense of a dot should be sufficient to determine if it's encoded or not.

Reply:
It is when you take https://github.com/secdev/scapy/pull/1968#discussion_r271718311 into account: the dot is only added on strings/unicodes.

Reply:
Okay, I thought the use case for passing bytes was to allow an already-encoded rrname. If that is not it, what is the use case for passing bytes? Just avoid adding the final dot but still do the encoding? But in that case, the encoding *will* still add the final dot, so what is the point?

Reply:
I see your point.
It’s a tricky situation because:

| Should be encoded | shouldn't |
| --- | --- |
| `b"data"` | `b"\x05[...]\x00"` |
| `b"anthingwithadot."` | `b"[...]\xc0\x02"` |
| `"anything"` or `u"anything"` | |

I should probably correct myself: we should only ignore correctly encoded bytes, even though they lead to decompression loops. Scapy converts strings to bytes, so any string check must be done prior to encoding. We can't make it accept all bytes and encode only string.

Reply:
The problem is of course that it's almost impossible to distinguish an encoded bytes from an unencoded bytes without dot. The `((x[-1] == b"\x00") or ((orb(x[-2]) & 0xc0)) == 0xc0)` checks one aspect of correct encoding, but it seems a little arbitrary to me. It only checks the final label, not any of the other ones - which risks yielding false positives. You could just as well check the first bytes, e.g. `(orb(x[0]) <= 63) or (orb(x[0]) & 0xc0 == 0xc0 and len(x) == 2)`. In addition, checking for correct encoding will make it more difficult to intentionally create packets with incorrect encoding.

So to me, it would make more sense that for a bytes, the distinction between already-encoded and non-encoded is whether it contains a dot or not. That can be achieved by leaving the check in any2i like you have it (i.e. not my alternative proposal) and removing the incomplete `((x[-1] == b"\x00") or ((orb(x[-2]) & 0xc0)) == 0xc0)`  check.

Reply:
Originally, Scapy didn't support passing the already-encoded value as-is. It was partially added so that a packet could be built in `dns_compress` with special pointers, without being re-I coded afterwards. (I did it). It wasn't intended to support all cases :/

A dot isn't enough: `b"test"` does not contain any but still should be encoded :/ This arbitrary test at least allows not to break any backward compatibility.